cmd/fleet/cron.go

```diff
@@ -0,0 +1,347 @@
+				continue
+			}
+		}
+
+		// Reread app config to be able to read latest data used by the webhook
+		// and update the ticker for the next run.
+		appConfig, err = ds.AppConfig(ctx)
+		if err != nil {
+			level.Error(logger).Log("config", "couldn't read app config", "err", err)
+			sentry.CaptureException(err)
+		} else {
+			ticker.Reset(appConfig.WebhookSettings.Interval.ValueOr(24 * time.Hour))
+			start = time.Now()
+		}
+
+		// We set the db lock durations to match the intervalReload.
+		maybeTriggerHostStatus(ctx, ds, logger, identifier, appConfig, intervalReload)
+		maybeTriggerFailingPoliciesWebhook(ctx, ds, logger, identifier, appConfig, intervalReload, failingPoliciesSet)
+
+		level.Debug(logger).Log("loop", "done")
+	}
+}
+
+func maybeTriggerHostStatus(
+	ctx context.Context,
+	ds fleet.Datastore,
+	logger kitlog.Logger,
+	identifier string,
+	appConfig *fleet.AppConfig,
+	lockDuration time.Duration,
+) {
+	if locked, err := ds.Lock(ctx, lockKeyWebhooksHostStatus, identifier, lockDuration); err != nil || !locked {
+		level.Debug(logger).Log("leader-host-status", "Not the leader. Skipping...")
+		return
+	}
+
+	if err := webhooks.TriggerHostStatusWebhook(
+		ctx, ds, kitlog.With(logger, "webhook", "host_status"), appConfig,
+	); err != nil {
+		level.Error(logger).Log("err", "triggering host status webhook", "details", err)
+		sentry.CaptureException(err)
+	}
+}
+
+func maybeTriggerFailingPoliciesWebhook(
+	ctx context.Context,
+	ds fleet.Datastore,
+	logger kitlog.Logger,
+	identifier string,
+	appConfig *fleet.AppConfig,
+	lockDuration time.Duration,
+	failingPoliciesSet fleet.FailingPolicySet,
+) {
+	if locked, err := ds.Lock(ctx, lockKeyWebhooksFailingPolicies, identifier, lockDuration); err != nil || !locked {
+		level.Debug(logger).Log("leader-failing-policies", "Not the leader. Skipping...")
+		return
+	}
+
+	if err := webhooks.TriggerFailingPoliciesWebhook(
+		ctx, ds, kitlog.With(logger, "webhook", "failing_policies"), appConfig, failingPoliciesSet, time.Now(),
+	); err != nil {
+		level.Error(logger).Log("err", "triggering failing policies webhook", "details", err)
+		sentry.CaptureException(err)
+	}
+}
+
+func cronWorker(ctx context.Context, ds fleet.Datastore, logger kitlog.Logger, identifier string) {
+	w := worker.NewWorker(ds, logger)
+
+	jira := worker.NewJira(ds, logger)
+	w.Register(jira)
+
+	logger = kitlog.With(logger, "key", lockKeyWorker)
+
+	ticker := time.NewTicker(10 * time.Second)
+	for {
+		select {
+		case <-ticker.C:
+			level.Debug(logger).Log("waiting", "done")
+			ticker.Reset(10 * time.Minute)
+		case <-ctx.Done():
+			level.Debug(logger).Log("exit", "done with cron.")
+			return
+		}
+
+		if locked, err := ds.Lock(ctx, lockKeyWorker, identifier, 10*time.Minute); err != nil {
+			level.Error(logger).Log("msg", "Error acquiring lock", "err", err)
+			continue
+		} else if !locked {
+			level.Debug(logger).Log("msg", "Not the leader. Skipping...")
+			continue
+		}
```

Comment:
:+1: yeah that looks good to me, more detailed info in the logs.

Reply:
Okay, I'll update the other crons to match